File: benchmarks/benchmark.py
#!/usr/bin/env python3
import numpy as np
import pandas as pd
from Bio import SeqIO
from datetime import datetime
from ahocorapy.keywordtree import KeywordTree


for n_markers in (1_000, 10_000, 100_000, 1_000_000,):
    markers = pd.read_csv("./markers.csv", header=None, index_col=0, nrows=n_markers)

    print(f"Number of markers: {n_markers}")

    trie = KeywordTree()
    start = datetime.now()
    for marker in markers.values.flatten():
        trie.add(marker)
    trie.finalize()
    end = datetime.now()
    print(f"Trie built in {(end-start).total_seconds():.1f} seconds")

    times = []
    for f in SeqIO.parse("./pseudo10001.fasta", "fasta"):
        start = datetime.now()
        res = trie.search_all(f.seq)
        n_matches = len(list(res))
        end = datetime.now()
        times.append((end - start).total_seconds())
        print(f"{f.id} : {n_matches:6} matches : {times[-1]:5.1f} seconds")
    
    print(f"Average matching time: {np.mean(times):5.1f} seconds")
    print("")

File: benchmarks/benchmark.cpp
#include <algorithm>
#include <iomanip>
#include <iostream>
#include <string>

#include <utils.hpp>
#include <aho_corasick.hpp>


void test_build_and_match(size_t num_markers) {
    auto markers = read_markers("../../data/markers.csv", num_markers);
    auto sequences = read_fasta_archive("../../data/pseudo10001.fasta.gz");

    aho_corasick::trie trie;
    std::cout << "Number of markers: " << num_markers << std::endl;
    auto t1 = Time::now();
    for (auto& m : markers)
        trie.insert(m._marker);
    auto t2 = Time::now();
    std::cout << "Trie built in " << Time::diff(t1, t2) << " seconds" << std::endl;

    for (const auto &s : sequences) {
        auto t3 = Time::now();
        auto result = trie.parse_text(s._sequence);
        auto t4 = Time::now();
        auto it = std::unique(result.begin(), result.end());
        std::cout
                << s._id
                << " matched in " << Time::diff(t3, t4) << " seconds, "
                << "found " << std::distance(result.begin(), it) << " matches." <<  std::endl;
    }

    std::cout << std::endl;
}


int main() {
    for (size_t n : {1000, 10000, 100000, 1000000})
        test_build_and_match(n);
    return 0;
}